### 🔹  Step 1: Importing Libraries, SQL Query, and Database Connection

In [87]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier

In [88]:
pip install imbalanced-learn

In [89]:
# Importer les bibliothèques nécessaires
import pandas as pd
import numpy as np
from sqlalchemy import create_engine


# Configurer la connexion à la base de données
# Remplacez les valeurs par vos propres informations de connexion
db_connection_string = 'mssql+pyodbc://sa:sasa@DESKTOP-CH9FCFH/DW_supplyChainProject?driver=ODBC+Driver+17+for+SQL+Server'
engine = create_engine(db_connection_string)

# Charger les données
fact_distribution_sales = pd.read_sql_query("""
    SELECT *
    FROM [DW_supplyChainProject].[dbo].[Fact_Distribution_Sales]
""", engine)

fact_production_storage = pd.read_sql_query("""
    SELECT *
    FROM [DW_supplyChainProject].[dbo].[Fact_Production_Storage]
""", engine)

dim_date = pd.read_sql_query("""
    SELECT *
    FROM [DW_supplyChainProject].[dbo].[Dim_Date]
""", engine)

dim_geography = pd.read_sql_query("""
    SELECT *
    FROM [DW_supplyChainProject].[dbo].[Dim_Geography]
""", engine)

dim_product = pd.read_sql_query("""
    SELECT *
    FROM [DW_supplyChainProject].[dbo].[Dim_Product]
""", engine)

# Afficher les premières lignes de chaque DataFrame pour vérification
print(fact_distribution_sales.head())
print(fact_production_storage.head())
print(dim_date.head())
print(dim_geography.head())
print(dim_product.head())

   Shop_Fk  Distributor_Fk  Product_Fk  Invoice_Fk  Invoice_Date_Fk  Order_Fk  \
0       41               1         491         375             1756    3563.0   
1       41               1         494         320             1518    2320.0   
2       41               1         498         147             1116       NaN   
3       41               1         504          28             1199    2095.0   
4       41               1         509         261              963       NaN   

   Order_Date_Fk  Distributor_Geography_Fk  Shop_Geography_Fk  \
0         1289.0                         1                  7   
1          956.0                         1                  7   
2            NaN                         1                  7   
3         1044.0                         1                  7   
4            NaN                         1                  7   

   Product_Quantity_Sold  Product_Unit_Price  Total_General  \
0                      7          140.300003   12164.160156

### 🔹Step 1:Data Preparation & Cleaning

In [91]:
# Vérifier les valeurs manquantes dans les DataFrames
print("Valeurs manquantes dans Fact_Distribution_Sales:")
print(fact_distribution_sales.isnull().sum())

print("\nValeurs manquantes dans Fact_Production_Storage:")
print(fact_production_storage.isnull().sum())

Valeurs manquantes dans Fact_Distribution_Sales:
Shop_Fk                        0
Distributor_Fk                 0
Product_Fk                     0
Invoice_Fk                     0
Invoice_Date_Fk                0
Order_Fk                    3909
Order_Date_Fk               3909
Distributor_Geography_Fk       0
Shop_Geography_Fk              0
Product_Quantity_Sold          0
Product_Unit_Price             0
Total_General                  0
Product_Quantity_Ordered     403
dtype: int64

Valeurs manquantes dans Fact_Production_Storage:
Warehouse_Fk                             0
Geography_Fk                             0
Product_Fk                               0
Product_Entry_Date_Fk                    0
Product_Exit_Date_Fk                     0
Manufacture_Date_Fk                      0
Expiry_Date_Fk                           0
Order_Fk                                 0
Order_Date_Fk                            0
Warehouse_Capacity                       0
Material_Quantity_Used       

In [92]:
print("Colonnes de Fact_Distribution_Sales:")
print(fact_distribution_sales.columns)

print("\nColonnes de Fact_Production_Storage:")
print(fact_production_storage.columns)

print("\nColonnes de Dim_Date:")
print(dim_date.columns)

print("\nColonnes de Dim_Geography:")
print(dim_geography.columns)

print("\nColonnes de Dim_Product:")
print(dim_product.columns)

Colonnes de Fact_Distribution_Sales:
Index(['Shop_Fk', 'Distributor_Fk', 'Product_Fk', 'Invoice_Fk',
       'Invoice_Date_Fk', 'Order_Fk', 'Order_Date_Fk',
       'Distributor_Geography_Fk', 'Shop_Geography_Fk',
       'Product_Quantity_Sold', 'Product_Unit_Price', 'Total_General',
       'Product_Quantity_Ordered'],
      dtype='object')

Colonnes de Fact_Production_Storage:
Index(['Warehouse_Fk', 'Geography_Fk', 'Product_Fk', 'Product_Entry_Date_Fk',
       'Product_Exit_Date_Fk', 'Manufacture_Date_Fk', 'Expiry_Date_Fk',
       'Order_Fk', 'Order_Date_Fk', 'Warehouse_Capacity',
       'Material_Quantity_Used', 'Product_Stored_Quantity',
       'Product_Production_Duration_Hours', 'Product_Quantity_Produced',
       'Product_Quantity_Ordered'],
      dtype='object')

Colonnes de Dim_Date:
Index(['Date_Pk', 'Date', 'Year', 'Month', 'Month_Name', 'Quarter', 'Semester',
       'Day', 'Weekday_Name', 'Weekday_Number', 'Is_Weekend'],
      dtype='object')

Colonnes de Dim_Geography:
Index(

In [93]:
print("Types de données dans Fact_Distribution_Sales:")
print(fact_distribution_sales.dtypes)

print("\nTypes de données dans Fact_Production_Storage:")
print(fact_production_storage.dtypes)

Types de données dans Fact_Distribution_Sales:
Shop_Fk                       int64
Distributor_Fk                int64
Product_Fk                    int64
Invoice_Fk                    int64
Invoice_Date_Fk               int64
Order_Fk                    float64
Order_Date_Fk               float64
Distributor_Geography_Fk      int64
Shop_Geography_Fk             int64
Product_Quantity_Sold         int64
Product_Unit_Price          float64
Total_General               float64
Product_Quantity_Ordered    float64
dtype: object

Types de données dans Fact_Production_Storage:
Warehouse_Fk                           int64
Geography_Fk                           int64
Product_Fk                             int64
Product_Entry_Date_Fk                  int64
Product_Exit_Date_Fk                   int64
Manufacture_Date_Fk                    int64
Expiry_Date_Fk                         int64
Order_Fk                               int64
Order_Date_Fk                          int64
Warehouse_Capacity

In [94]:
# Afficher les valeurs uniques de Product_Fk
unique_distribution_products = fact_distribution_sales['Product_Fk'].unique()
unique_production_products = fact_production_storage['Product_Fk'].unique()

print("Valeurs uniques de Product_Fk dans Fact_Distribution_Sales:")
print(unique_distribution_products)

print("\nValeurs uniques de Product_Fk dans Fact_Production_Storage:")
print(unique_production_products)

# Optionnel : Vérifier les produits présents dans un DataFrame mais pas dans l'autre
missing_in_production = set(unique_distribution_products) - set(unique_production_products)
missing_in_distribution = set(unique_production_products) - set(unique_distribution_products)

print("\nProduits manquants dans Fact_Production_Storage:")
print(missing_in_production)

print("\nProduits manquants dans Fact_Distribution_Sales:")
print(missing_in_distribution)


Valeurs uniques de Product_Fk dans Fact_Distribution_Sales:
[ 491  494  498 ... 2086 3498 5013]

Valeurs uniques de Product_Fk dans Fact_Production_Storage:
[ 491  494  498 ... 5143 5150 5154]

Produits manquants dans Fact_Production_Storage:
{np.int64(2690), np.int64(2053), np.int64(519), np.int64(776), np.int64(905), np.int64(3339), np.int64(4749), np.int64(1549), np.int64(1808), np.int64(3350), np.int64(1430), np.int64(1814), np.int64(2841), np.int64(3482), np.int64(5147), np.int64(1946), np.int64(2207), np.int64(1823), np.int64(545), np.int64(1057), np.int64(2339), np.int64(1185), np.int64(3364), np.int64(3620), np.int64(3878), np.int64(681), np.int64(3498), np.int64(4525), np.int64(3757), np.int64(3631), np.int64(1711), np.int64(2225), np.int64(3888), np.int64(3508), np.int64(4917), np.int64(4024), np.int64(4666), np.int64(3389), np.int64(2878), np.int64(2496), np.int64(4551), np.int64(3143), np.int64(1484), np.int64(845), np.int64(590), np.int64(3278), np.int64(2259), np.int64(29

#### 🔹 Step 3: Grouping and Aggregating Data

In [96]:
# Fusionner les DataFrames avec une fusion externe
merged_data = fact_distribution_sales.merge(
    fact_production_storage, on='Product_Fk', suffixes=('_sales', '_storage'), how='inner'
).merge(
    dim_date, left_on='Invoice_Date_Fk', right_on='Date_Pk', how='inner'
).merge(
    dim_geography, left_on='Distributor_Geography_Fk', right_on='Geography_Pk', how='inner'
).merge(
    dim_product, left_on='Product_Fk', right_on='Product_Pk', how='inner'
)

In [97]:
# Vérifier les valeurs manquantes dans merged_data
missing_values = merged_data.isnull().sum()
print("Valeurs manquantes dans merged_data:")
print(missing_values[missing_values > 0])

Valeurs manquantes dans merged_data:
Order_Fk_sales                       140240
Order_Date_Fk_sales                  140240
Product_Production_Duration_Hours    248480
Product_Quantity_Produced            248480
dtype: int64


In [98]:
nombre_de_lignes = merged_data.shape[0]
print(f"Nombre de lignes dans merged_data : {nombre_de_lignes}")

Nombre de lignes dans merged_data : 285520


In [99]:
# Supprimer les lignes avec des valeurs nulles dans les colonnes spécifiées
merged_data.dropna(subset=[
    'Product_Quantity_Produced',
    'Order_Fk_sales',
    'Order_Date_Fk_sales',
    'Product_Production_Duration_Hours'
], inplace=True)

# Vérifier le nombre de lignes après la suppression
print(f"Nombre de lignes après suppression : {len(merged_data)}")

Nombre de lignes après suppression : 18320


In [100]:
# Vérifier les valeurs manquantes dans merged_data
missing_values = merged_data.isnull().sum()
print("Valeurs manquantes dans merged_data:")
print(missing_values[missing_values > 0])

Valeurs manquantes dans merged_data:
Series([], dtype: int64)


   ### 🔹Step4: Create the target variable and select features
   

In [102]:
# Calculer la médiane des quantités stockées, en ignorant les valeurs manquantes
median_stock_quantity = merged_data['Product_Stored_Quantity'].median()

# Créer la variable cible : si le produit est en rupture de stock
merged_data['is_out_of_stock'] = (merged_data['Product_Stored_Quantity'].fillna(0) < median_stock_quantity).astype(int)

In [103]:
    # Sélectionner les caractéristiques (features) et la cible (target)
    features = merged_data[['Product_Quantity_Sold', 'Product_Unit_Price', 
                             'Product_Quantity_Ordered_storage', 
                             'Warehouse_Capacity',
                             'Product_Production_Duration_Hours']]
    target = merged_data['is_out_of_stock']

In [104]:
# Encodage des variables catégorielles
features = pd.get_dummies(features, drop_first=True)

# Afficher les premières lignes du DataFrame prétraité
print("\nCaractéristiques prétraitées:")
print(features.head())

print("\nCible prétraitée:")
print(target.head())


Caractéristiques prétraitées:
   Product_Quantity_Sold  Product_Unit_Price  \
0                      7          140.300003   
1                      7          140.300003   
2                      7          140.300003   
3                      7          140.300003   
4                      7          140.300003   

   Product_Quantity_Ordered_storage  Warehouse_Capacity  \
0                                 6                 500   
1                                 6                 300   
2                                 6                 400   
3                                 6                 250   
4                                 6                 350   

   Product_Production_Duration_Hours  
0                               24.0  
1                               24.0  
2                               24.0  
3                               24.0  
4                               24.0  

Cible prétraitée:
0    1
1    0
2    0
3    1
4    0
Name: is_out_of_stock, dtype: int64


In [105]:
# Division des données
from imblearn.over_sampling import SMOTE

X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.2, random_state=42)
smote = SMOTE(random_state=42)
X_train_res, y_train_res = smote.fit_resample(X_train, y_train)

C:\Users\oumeyma\AppData\Roaming\Python\Python312\site-packages\sklearn\base.py:474: FutureWarning: `BaseEstimator._validate_data` is deprecated in 1.6 and will be removed in 1.7. Use `sklearn.utils.validation.validate_data` instead. This function becomes public and is part of the scikit-learn developer API.
  warnings.warn(


   ### 🔹Step 5: Apply Grid serach 


In [107]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'n_estimators': [100, 200],
    'max_depth': [10, 20, None],
    'min_samples_split': [2, 5]
}

grid_search = GridSearchCV(RandomForestClassifier(random_state=42), param_grid, cv=5, scoring='f1')
grid_search.fit(X_train_res, y_train_res)
print("Meilleurs paramètres :", grid_search.best_params_)

Meilleurs paramètres : {'max_depth': 20, 'min_samples_split': 5, 'n_estimators': 100}


   ### 🔹Step 6: Apply Algo RandomForestClassifier


In [109]:
from sklearn.metrics import classification_report, confusion_matrix

# Créer et entraîner le modèle
model = RandomForestClassifier(n_estimators=100, max_depth=20, random_state=42,min_samples_split= 5)
model.fit(X_train_res, y_train_res)

# Faire des prédictions
y_pred = model.predict(X_test)

# Évaluer le modèle
print("\nRapport de classification:")
print(classification_report(y_test, y_pred))
confusion = confusion_matrix(y_test, y_pred)
print("Matrice de confusion:\n", confusion)


Rapport de classification:
              precision    recall  f1-score   support

           0       0.63      0.64      0.64      1842
           1       0.63      0.62      0.63      1822

    accuracy                           0.63      3664
   macro avg       0.63      0.63      0.63      3664
weighted avg       0.63      0.63      0.63      3664

Matrice de confusion:
 [[1186  656]
 [ 690 1132]]


   ### 🔹Step 7: Apply Grid serach for XgBoost


In [111]:
from sklearn.model_selection import GridSearchCV
from xgboost import XGBClassifier

param_grid = {
    'n_estimators': [100, 200],
    'max_depth': [3, 5, 7],
    'learning_rate': [0.01, 0.1, 0.2],
    'subsample': [0.8, 1.0]
}

grid_search = GridSearchCV(XGBClassifier(use_label_encoder=False, eval_metric='logloss', random_state=42), 
                           param_grid, cv=5, scoring='f1')
grid_search.fit(X_train_res, y_train_res)
print("Meilleurs paramètres :", grid_search.best_params_)

C:\Users\oumeyma\AppData\Roaming\Python\Python312\site-packages\xgboost\training.py:183: UserWarning: [22:22:01] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
C:\Users\oumeyma\AppData\Roaming\Python\Python312\site-packages\xgboost\training.py:183: UserWarning: [22:22:01] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
C:\Users\oumeyma\AppData\Roaming\Python\Python312\site-packages\xgboost\training.py:183: UserWarning: [22:22:01] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
C:\Users\oumeyma\AppData\Roaming\Python\Python312\site-packages\xgboost\training.py:183: UserWarning: [22:22:01] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\

Meilleurs paramètres : {'learning_rate': 0.2, 'max_depth': 7, 'n_estimators': 200, 'subsample': 1.0}


   ### 🔹Step 8: Apply Smot and Algo XgBoost

In [113]:
from sklearn.metrics import classification_report, confusion_matrix
from xgboost import XGBClassifier
from imblearn.over_sampling import SMOTE
# Appliquer SMOTE pour équilibrer les classes
smote = SMOTE(random_state=42)
X_train_res, y_train_res = smote.fit_resample(X_train, y_train)

# Créer et entraîner le modèle XGBoost avec les meilleurs paramètres
#model = XGBClassifier(learning_rate=0.2, max_depth=7, n_estimators=200, subsample=1.0, 
 #                     eval_metric='logloss', random_state=42)
#model.fit(X_train_res, y_train_res)

# Créer et entraîner le modèle XGBoost
model = XGBClassifier(use_label_encoder=False, eval_metric='logloss', random_state=42)
model.fit(X_train_res, y_train_res)

# Faire des prédictions
y_pred = model.predict(X_test)

# Évaluer le modèle
print("\nRapport de classification:")
print(classification_report(y_test, y_pred))
confusion = confusion_matrix(y_test, y_pred)
print("Matrice de confusion:\n", confusion)


C:\Users\oumeyma\AppData\Roaming\Python\Python312\site-packages\sklearn\base.py:474: FutureWarning: `BaseEstimator._validate_data` is deprecated in 1.6 and will be removed in 1.7. Use `sklearn.utils.validation.validate_data` instead. This function becomes public and is part of the scikit-learn developer API.
  warnings.warn(
C:\Users\oumeyma\AppData\Roaming\Python\Python312\site-packages\xgboost\training.py:183: UserWarning: [22:22:45] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)



Rapport de classification:
              precision    recall  f1-score   support

           0       0.69      0.70      0.69      1842
           1       0.69      0.68      0.68      1822

    accuracy                           0.69      3664
   macro avg       0.69      0.69      0.69      3664
weighted avg       0.69      0.69      0.69      3664

Matrice de confusion:
 [[1286  556]
 [ 585 1237]]


✅ True Negatives (TN): 1286 (no break correctly predicted)  
✅ False Positives (FP): 556 (no break predicted as a break)  
✅ False Negatives (FN): 585 (break predicted as no break)  
✅ True Positives (TP): 1237 (break correctly predicted)  

#### ✅  Vrais Négatifs (VN) : 1286 (non rupture correctement prédite)    
#### ✅   Faux Positifs (FP) : 556 (non rupture prédite comme rupture)  
#### ✅  Faux Négatifs (FN) : 585 (rupture prédite comme non rupture)  
#### ✅  Vrais Positifs (VP) : 1237 (rupture correctement prédite)  

In [115]:
# Ajouter les colonnes utiles pour interprétation
X_test_reset = X_test.reset_index(drop=True)
y_test_reset = y_test.reset_index(drop=True)

# Récupérer les colonnes d'identification (ex: Product_Name) depuis merged_data
product_info = merged_data[['Product_Name', 'Product_Id']].iloc[X_test_reset.index].reset_index(drop=True)

# Créer le tableau de résultats
resultats = pd.concat([product_info, X_test_reset], axis=1)
resultats["Rupture réelle"] = y_test_reset

# Optionnel : rendre plus lisible
resultats["Rupture réelle"] = resultats["Rupture réelle"].map({0: "0", 1: "1"})

# Afficher les 10 premiers résultats
print(resultats.head(10))

# Sauvegarder en CSV si besoin
resultats.to_csv("tableau_predictions_rupture.csv", index=False)


   Product_Name Product_Id  Product_Quantity_Sold  Product_Unit_Price  \
0  Hair Serum 1     PR-001                      7           10.370000   
1  Hair Serum 1     PR-001                     10           98.190002   
2  Hair Serum 1     PR-001                     11           75.080002   
3  Hair Serum 1     PR-001                      6           46.150002   
4  Hair Serum 1     PR-001                     16           23.430000   
5  Hair Serum 1     PR-001                     15          103.980003   
6  Hair Serum 1     PR-001                     20          165.800003   
7  Hair Serum 1     PR-001                     10           46.150002   
8  Hair Serum 1     PR-001                     13          102.610001   
9  Hair Serum 1     PR-001                     12           83.349998   

   Product_Quantity_Ordered_storage  Warehouse_Capacity  \
0                                 1                 310   
1                                 9                 350   
2                  

In [116]:
import joblib

joblib.dump(model, 'xgb_classifier.pkl')


['xgb_classifier.pkl']